python/mujincontrollerclient/controllerclientbase.py

```diff
@@ -887,3 +887,29 @@ def SetITLProgram(self, programName, data, fields=None, usewebapi=True, timeout=
     def DeleteITLProgram(self, programName, usewebapi=True, timeout=5):
         assert(usewebapi)
         self._webclient.APICall('DELETE', u'itl/%s/' % programName, timeout=timeout)
+
+    #
+    # Backup related
+    #
+
+    def Backup(self, media=True, config=True, usewebapi=True, timeout=30, **kwargs):
+        assert(usewebapi)
+        params = {
+            'media': 'true' if media else 'false',
+            'config': 'true' if config else 'false',
+        }
+        params.update(kwargs)
+        response = self._webclient.Request('GET', '/backup/', stream=True, params=params, timeout=timeout)
+        if response.status_code != 200:
+            raise ControllerClientError(_('failed to download backup from controller, status code is: %d') % response.status_code)
+        return response
```

Comment:
my (unofficial) impl here is:

```
                response.raw.decode_content = True
                return response.raw
```

in this way you can improve _RunMain like this:

```
        with tarfile.open(fileobj=response, mode='r:*') as tar:
            config = _LoadConfigTar(tar)
```

I mean, I suppose stream=True should be used in this way.

Reply:
`decode_content = True` is for decoding transparent gzip encoding.

Reply:
> I mean, I suppose stream=True should be used in this way.

@ziyan how do you think?

Reply:
What if caller wants to save the tar gz as-is to disk?

Reply:
```
with open('backup.tar.gz','w') as f:
  shutil.copyfileobj(response,f)
```

Reply:
So what's the advantage?

Reply:
So I am confused as to what `response.raw.decode_content = True` does exactly? Who is decoding and how? Is is decoding with threads? Is it caching? Sometimes you need to ask yourself if the code has too much magic, and whether it's worth it to sacrifice the readability.

Reply:
response.raw is urllib3 handle: https://github.com/urllib3/urllib3/blob/master/src/urllib3/response.py#L492

it decodes transparent encoding (if you are not aware, you are returning tar.gz with recompressing "content-encoding: gzip").
when read() happens, urllib3 passes data to zlib.decompressobj without thread. this decompressobj caches partial readed data.

Reply:
`stream=True should be used in this way` meant you could access backup content without holding full content in memory. (sorry if I was not clear)

Reply:
@ziyan if we are not sure about API design, how about introducing `returnRawHandle` kwargs? adding kwargs is always possible (unlike C++)...